await-generator/src/SOFe/AwaitGenerator/PubSub.php

```diff
@@ -0,0 +1,106 @@
+<?php
+
+/*
+ * await-generator
+ *
+ * Copyright (C) 2023 SOFe
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+declare(strict_types=1);
+
+namespace SOFe\AwaitGenerator;
+
+use RuntimeException;
+
+/**
+ * A pubsub allows coroutines to publish a message that is received by all subscribres.
+ * @template T
+ */
+final class PubSub{
+	/** @var Channel<T>[] */
+	private array $subscribers = [];
+
+	/**
+	 * If a subscriber is lagging for more than $maxLag items behind,
+	 * a RuntimeException is thrown to help detect memory leaks in advance.
+	 * Set it to `null` to disable the check.
+	 */
+	public function __construct(
+		private ?int $maxLag = 10000,
+	) {}
+
+	/**
+	 * Publishes a message and return.
+	 *
+	 * This method does not block until the event is actually received by subscribers.
+	 *
+	 * @phpstan-param T $item
+	 */
+	public function publish($item) : void{
+		foreach($this->subscribers as $subscriber) {
+			$subscriber->sendWithoutWait($item);
+			if($this->maxLag !== null && $subscriber->getSendQueueSize() > $this->maxLag) {
+				throw new RuntimeException("A subscriber has been lagging for $this->maxLag items. Forgot to call \$traverser->interrupt()?");
+			}
+		}
+	}
+
+	/**
+	 * Subscribes to the messages of this topic.
+	 *
+	 * The returned traverser does not return any messages published prior to calling `subscribe()`.
+	 *
+	 * Subscribers are tracked in the `PubSub`.
+	 * To avoid memory leak,
+	 * callers to this method must interrupt the traverser in a `finally` block:
```

Comment:
```suggestion
	 * callers to this method must interrupt the traverser in a `finally` block:
	 * WARNING: Otherwise, a RuntimeException will be thrown once {@link $this->maxLag}
	 * is reached!
```

Reply:
`maxLag` is not a public field. Don't link a private field from a public documentation. What about just writing `maxLag`?

Reply:
```suggestion
	 * callers to this method must interrupt the traverser in a `finally` block:
	 * WARNING: Otherwise, a RuntimeException will be thrown once `maxLag` is reached!
```